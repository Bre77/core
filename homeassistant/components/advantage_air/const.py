"""Constants used by Advantage Air integration."""
DOMAIN = "advantage_air"
ADVANTAGE_AIR_AIRCONS = "aircons"
ADVANTAGE_AIR_LIGHTS = "myLights"
ADVANTAGE_AIR_THINGS = "myThings"
ADVANTAGE_AIR_RETRY = 10
ADVANTAGE_AIR_STATE_OPEN = "open"
ADVANTAGE_AIR_STATE_CLOSE = "close"
ADVANTAGE_AIR_STATE_ON = "on"
ADVANTAGE_AIR_STATE_OFF = "off"
ADVANTAGE_AIR_SET_AIRCON = "async_set_aircon"
ADVANTAGE_AIR_SET_LIGHT = "async_set_light"
ADVANTAGE_AIR_SET_THING = "async_set_thing"
ADVANTAGE_AIR_COORDINATOR = "coordinator"
