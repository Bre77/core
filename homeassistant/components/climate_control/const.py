"""Constants used by Climate Control integration."""
DOMAIN = "climate_control"
CONF_CLIMATE_ENTITY = "climate_entity"
CONF_AREAS = "areas"
CONF_ZONES = "zones"
CONF_COVER_ENTITY = "cover_entity"
CONF_SENSOR_ENTITY = "sensor_entity"
