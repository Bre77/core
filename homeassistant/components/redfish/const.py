"""Constants used by Refish integration."""
DOMAIN = "redfish"
REDFISH_POWER_ON = "On"
REDFISH_POWER_OFF = "Off"
