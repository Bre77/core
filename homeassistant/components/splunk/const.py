HEC_TIME = "time"
HEC_HOST = "host"
HEC_SOURCE = "source"
HEC_SOURCETYPE = "sourcetype"
HEC_INDEX = "index"
HEC_FIELDS = "fields"
HEC_EVENT = "event"
