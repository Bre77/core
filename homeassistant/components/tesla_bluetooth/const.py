"""Constants for the Tesla Bluetooth integration."""

import logging

DOMAIN = "tesla_bluetooth"
PRIVATE_KEY_FILE = "config/tesla_key.pem"
LOGGER = logging.getLogger(__package__)
MANUFACTURER_ID = 76
SERVICE_UUIDS = ["00001122-0000-1000-8000-00805f9b34fb"]

# UUIDs
GENERIC_ACCESS_SERVICE_UUID = "00001800-0000-1000-8000-00805f9b34fb"
DEVICE_NAME_UUID = "00002a00-0000-1000-8000-00805f9b34fb"
SEND_UUID = "00000212-b2d1-43f0-9b88-960cebf8b91e"
RECV_UUID = "00000213-b2d1-43f0-9b88-960cebf8b91e"
VERSION_UUID = "00000214-b2d1-43f0-9b88-960cebf8b91e"
