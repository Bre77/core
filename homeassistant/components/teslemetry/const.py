"""Constants used by Teslemetry integration."""
from __future__ import annotations

from enum import IntEnum, StrEnum

DOMAIN = "teslemetry"

MODELS = {
    "model3": "Model 3",
    "modelx": "Model X",
    "modely": "Model Y",
    "models": "Model S",
}


class TeslemetryStatus(StrEnum):
    """Teslemetry status."""

    ASLEEP = "asleep"
    ONLINE = "online"


class TeslemetrySeatHeaterOptions(StrEnum):
    """Teslemetry seat heater options."""

    OFF = "off"
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class TeslemetryClimateKeeper(StrEnum):
    """Teslemetry Climate Keeper Modes."""

    OFF = "off"
    ON = "on"
    DOG = "dog"
    CAMP = "camp"


class TeslemetryUpdateStatus(StrEnum):
    """Teslemetry Update Statuses."""

    AVAILABLE = "available"
    DOWNLOADING = "downloading"
    INSTALLING = "installing"
    WIFI_WAIT = "downloading_wifi_wait"
    SCHEDULED = "scheduled"


class TeslemetryCoverStates(IntEnum):
    """Teslemetry Cover states."""

    CLOSED = 0
    OPEN = 1
