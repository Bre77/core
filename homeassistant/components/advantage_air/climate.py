"""Climate platform for Advantage Air integration."""
from __future__ import annotations

import logging

from sqlalchemy import Float

from homeassistant.components.climate import ClimateEntity
from homeassistant.components.climate.const import (
    ATTR_TARGET_TEMP_HIGH,
    ATTR_TARGET_TEMP_LOW,
    FAN_AUTO,
    FAN_HIGH,
    FAN_LOW,
    FAN_MEDIUM,
    ClimateEntityFeature,
    HVACMode,
)
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import ATTR_TEMPERATURE, PRECISION_WHOLE, TEMP_CELSIUS
from homeassistant.core import HomeAssistant
from homeassistant.helpers import entity_platform
from homeassistant.helpers.entity_platform import AddEntitiesCallback

from .const import (
    ADVANTAGE_AIR_MYAUTO_ENABLE,
    ADVANTAGE_AIR_MYTEMP_ENABLE,
    ADVANTAGE_AIR_STATE_CLOSE,
    ADVANTAGE_AIR_STATE_OFF,
    ADVANTAGE_AIR_STATE_ON,
    ADVANTAGE_AIR_STATE_OPEN,
    DOMAIN as ADVANTAGE_AIR_DOMAIN,
)
from .entity import AdvantageAirEntity

ADVANTAGE_AIR_HVAC_MODES = {
    "heat": HVACMode.HEAT,
    "cool": HVACMode.COOL,
    "vent": HVACMode.FAN_ONLY,
    "dry": HVACMode.DRY,
    "myauto": HVACMode.AUTO,
}
HASS_HVAC_MODES = {v: k for k, v in ADVANTAGE_AIR_HVAC_MODES.items()}

ADVANTAGE_AIR_FAN_MODES = {
    "auto": FAN_AUTO,
    "low": FAN_LOW,
    "medium": FAN_MEDIUM,
    "high": FAN_HIGH,
}
HASS_FAN_MODES = {v: k for k, v in ADVANTAGE_AIR_FAN_MODES.items()}
FAN_SPEEDS = {FAN_LOW: 30, FAN_MEDIUM: 60, FAN_HIGH: 100}

ADVANTAGE_AIR_MYAUTO = "MyAuto"
ADVANTAGE_AIR_MYTEMP = "MyTemp"
ADVANTAGE_AIR_MYZONE = "MyZone"
ADVANTAGE_AIR_HEAT_TARGET = "myAutoHeatTargetTemp"
ADVANTAGE_AIR_COOL_TARGET = "myAutoCoolTargetTemp"
ADVANTAGE_AIR_SERVICE_SET_MYZONE = "set_myzone"

PARALLEL_UPDATES = 0

_LOGGER = logging.getLogger(__name__)


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up AdvantageAir climate platform."""

    instance = hass.data[ADVANTAGE_AIR_DOMAIN][config_entry.entry_id]

    entities: list[ClimateEntity] = []
    for ac_key, ac_device in instance["coordinator"].data["aircons"].items():
        entities.append(AdvantageAirAC(instance, ac_key))
        for zone_key, zone in ac_device["zones"].items():
            # Only add zone climate control when zone is in temperature control
            if zone["type"] != 0:
                entities.append(AdvantageAirZone(instance, ac_key, zone_key))
    async_add_entities(entities)

    platform = entity_platform.async_get_current_platform()
    platform.async_register_entity_service(
        ADVANTAGE_AIR_SERVICE_SET_MYZONE,
        {},
        "set_myzone",
    )


class AdvantageAirAC(AdvantageAirEntity, ClimateEntity):
    """AdvantageAir AC unit."""

    _attr_fan_modes = [FAN_LOW, FAN_MEDIUM, FAN_HIGH, FAN_AUTO]
    _attr_preset_modes = [ADVANTAGE_AIR_MYZONE, ADVANTAGE_AIR_MYTEMP]
    _attr_temperature_unit = TEMP_CELSIUS
    _attr_target_temperature_step = PRECISION_WHOLE
    _attr_max_temp = 32
    _attr_min_temp = 16

    def __init__(self, instance, ac_key):
        """Initialize an AdvantageAir AC unit."""
        super().__init__(instance, ac_key)
        self._attr_unique_id = f'{self.coordinator.data["system"]["rid"]}-{ac_key}'

        self._attr_hvac_modes = [
            HVACMode.OFF,
            HVACMode.COOL,
            HVACMode.HEAT,
            HVACMode.FAN_ONLY,
            HVACMode.DRY,
        ]
        # Add "auto" HVAC mode only if the system supports it
        if ADVANTAGE_AIR_MYAUTO_ENABLE in self._ac:
            self._attr_hvac_modes += [HVACMode.AUTO]

        self._attr_supported_features = ClimateEntityFeature.FAN_MODE
        # Add "MyTemp" preset and the preset feature only if the system supports it
        if ADVANTAGE_AIR_MYTEMP_ENABLE in self._ac:
            self._attr_supported_features |= ClimateEntityFeature.PRESET_MODE

    @property
    def target_temperature(self) -> Float:
        """Return the current target temperature."""
        return self._ac["setTemp"]

    @property
    def hvac_mode(self):
        """Return the current HVAC modes."""
        if self._ac["state"] == ADVANTAGE_AIR_STATE_ON:
            return ADVANTAGE_AIR_HVAC_MODES.get(self._ac["mode"])
        return HVACMode.OFF

    @property
    def fan_mode(self):
        """Return the current fan modes."""
        return ADVANTAGE_AIR_FAN_MODES.get(self._ac["fan"])

    @property
    def preset_mode(self):
        """Return the current preset mode."""
        if self._ac.get(ADVANTAGE_AIR_MYAUTO_ENABLE):
            return ADVANTAGE_AIR_MYAUTO
        if self._ac.get(ADVANTAGE_AIR_MYTEMP_ENABLE):
            return ADVANTAGE_AIR_MYTEMP
        return ADVANTAGE_AIR_MYZONE

    @property
    def supported_features(self) -> int:
        """Return the list of supported features."""
        if self.hvac_mode == HVACMode.AUTO:
            return (
                self._attr_supported_features
                | ClimateEntityFeature.TARGET_TEMPERATURE_RANGE
            )
        return self._attr_supported_features | ClimateEntityFeature.TARGET_TEMPERATURE

    @property
    def target_temperature_high(self) -> float | None:
        """Return the temperature cool mode is enabled."""
        return self._ac.get(ADVANTAGE_AIR_COOL_TARGET, 24)

    @property
    def target_temperature_low(self) -> float | None:
        """Return the temperature heat mode is enabled."""
        return self._ac.get(ADVANTAGE_AIR_HEAT_TARGET, 20)

    async def async_set_hvac_mode(self, hvac_mode):
        """Set the HVAC Mode and State."""
        if hvac_mode == HVACMode.OFF:
            await self.async_change(
                {self.ac_key: {"info": {"state": ADVANTAGE_AIR_STATE_OFF}}}
            )
        else:
            await self.async_change(
                {
                    self.ac_key: {
                        "info": {
                            "state": ADVANTAGE_AIR_STATE_ON,
                            "mode": HASS_HVAC_MODES.get(hvac_mode),
                        }
                    }
                }
            )

    async def async_set_fan_mode(self, fan_mode):
        """Set the Fan Mode."""
        await self.async_change(
            {self.ac_key: {"info": {"fan": HASS_FAN_MODES.get(fan_mode)}}}
        )

    async def async_set_temperature(self, **kwargs):
        """Set the Temperature."""
        if ATTR_TEMPERATURE in kwargs:
            await self.async_change(
                {self.ac_key: {"info": {"setTemp": kwargs[ATTR_TEMPERATURE]}}}
            )
        if ATTR_TARGET_TEMP_LOW in kwargs and ATTR_TARGET_TEMP_HIGH in kwargs:
            await self.async_change(
                {
                    self.ac_key: {
                        "info": {
                            ADVANTAGE_AIR_COOL_TARGET: kwargs[ATTR_TARGET_TEMP_HIGH],
                            ADVANTAGE_AIR_HEAT_TARGET: kwargs[ATTR_TARGET_TEMP_LOW],
                        }
                    }
                }
            )

    async def async_set_preset_mode(self, preset_mode: str) -> None:
        """Set the preset mode."""
        await self.async_change(
            {
                self.ac_key: {
                    "info": {
                        ADVANTAGE_AIR_MYTEMP_ENABLE: preset_mode
                        == ADVANTAGE_AIR_MYTEMP,
                    }
                }
            }
        )


class AdvantageAirZone(AdvantageAirEntity, ClimateEntity):
    """AdvantageAir Zone control."""

    _attr_hvac_modes = [HVACMode.OFF, HVACMode.HEAT_COOL]
    _attr_supported_features = ClimateEntityFeature.TARGET_TEMPERATURE
    _attr_temperature_unit = TEMP_CELSIUS
    _attr_target_temperature_step = PRECISION_WHOLE
    _attr_max_temp = 32
    _attr_min_temp = 16

    def __init__(self, instance, ac_key, zone_key):
        """Initialize an AdvantageAir Zone control."""
        super().__init__(instance, ac_key, zone_key)
        self._attr_name = self._zone["name"]
        self._attr_unique_id = (
            f'{self.coordinator.data["system"]["rid"]}-{ac_key}-{zone_key}'
        )

    @property
    def hvac_mode(self) -> HVACMode:
        """Return the current state as HVAC mode."""
        if self._zone["state"] == ADVANTAGE_AIR_STATE_OPEN:
            return HVACMode.HEAT_COOL
        return HVACMode.OFF

    @property
    def current_temperature(self) -> float | None:
        """Return the current temperature."""
        return self._zone["measuredTemp"]

    @property
    def target_temperature(self) -> float:
        """Return the target temperature."""
        return self._zone["setTemp"]

    async def async_set_hvac_mode(self, hvac_mode) -> None:
        """Set the HVAC Mode and State."""
        await self.async_change(
            {
                self.ac_key: {
                    "zones": {
                        self.zone_key: {
                            "state": (hvac_mode == HVACMode.OFF)[
                                ADVANTAGE_AIR_STATE_OPEN, ADVANTAGE_AIR_STATE_CLOSE
                            ]
                        }
                    }
                }
            }
        )

    async def async_set_temperature(self, **kwargs) -> None:
        """Set the Temperature."""
        temp = kwargs.get(ATTR_TEMPERATURE)
        await self.async_change(
            {self.ac_key: {"zones": {self.zone_key: {"setTemp": temp}}}}
        )

    async def set_myzone(self, **kwargs):
        """Set this zone as the 'MyZone'."""
        _LOGGER.warning(
            "The advantage_air.set_myzone service has been deprecated and will be removed in a future version, please use the select.select_option service on the MyZone entity"
        )
        await self.async_change(
            {self.ac_key: {"info": {"myZone": self._zone["number"]}}}
        )
