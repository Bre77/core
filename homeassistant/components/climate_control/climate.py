"""Climate platform for Climate Control integration."""
from homeassistant.components.climate import ClimateEntity

from .const import DOMAIN


async def async_setup_entry(hass, config_entry, async_add_entities):
    """Set up ClimateControl climate platform."""

    instance = hass.data[ADVANTAGE_AIR_DOMAIN][config_entry.entry_id]

    entities = []
    for ac_key, ac_device in instance["coordinator"].data["aircons"].items():
        entities.append(ClimateControlAC(instance, ac_key))
        for zone_key, zone in ac_device["zones"].items():
            # Only add zone climate control when zone is in temperature control
            if zone["type"] != 0:
                entities.append(ClimateControlZone(instance, ac_key, zone_key))
    async_add_entities(entities)

    platform = entity_platform.async_get_current_platform()
    platform.async_register_entity_service(
        ADVANTAGE_AIR_SERVICE_SET_MYZONE,
        {},
        "set_myzone",
    )


class ClimateControlClimateEntity(ClimateControlEntity, ClimateEntity):