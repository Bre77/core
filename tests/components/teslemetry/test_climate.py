"""Test the teslemetry climate platform."""

from syrupy import SnapshotAssertion
from tesla_fleet_api.exceptions import TeslaFleetError

from homeassistant.components.climate import (
    ATTR_HVAC_MODE,
    ATTR_PRESET_MODE,
    ATTR_TEMPERATURE,
    DOMAIN as CLIMATE_DOMAIN,
    SERVICE_SET_HVAC_MODE,
    SERVICE_SET_PRESET_MODE,
    SERVICE_SET_TEMPERATURE,
    SERVICE_TURN_ON,
    HVACMode,
)
from homeassistant.const import ATTR_ENTITY_ID
from homeassistant.core import HomeAssistant

from tests.common import MockConfigEntry


async def test_climate(
    hass: HomeAssistant,
    config_entry_mock: MockConfigEntry,
    snapshot: SnapshotAssertion,
    teslemetry_vehicle_specific_mock,
) -> None:
    """Tests that the climate entity is correct."""

    config_entry_mock.add_to_hass(hass)
    assert len(hass.states.async_all(CLIMATE_DOMAIN)) == snapshot(name="all")

    entity_id = "climate.test_climate"
    assert hass.states.get(entity_id) == snapshot

    await hass.services.async_call(
        CLIMATE_DOMAIN,
        SERVICE_SET_HVAC_MODE,
        {ATTR_ENTITY_ID: [entity_id], ATTR_HVAC_MODE: HVACMode.HEAT_COOL},
        blocking=True,
    )
    assert hass.states.get(entity_id) == snapshot

    await hass.services.async_call(
        CLIMATE_DOMAIN,
        SERVICE_SET_TEMPERATURE,
        {ATTR_ENTITY_ID: [entity_id], ATTR_TEMPERATURE: 20.5},
        blocking=True,
    )
    assert hass.states.get(entity_id) == snapshot

    await hass.services.async_call(
        CLIMATE_DOMAIN,
        SERVICE_SET_PRESET_MODE,
        {ATTR_ENTITY_ID: [entity_id], ATTR_PRESET_MODE: "keep"},
        blocking=True,
    )
    assert hass.states.get(entity_id) == snapshot

    await hass.services.async_call(
        CLIMATE_DOMAIN,
        SERVICE_SET_HVAC_MODE,
        {ATTR_ENTITY_ID: [entity_id], ATTR_HVAC_MODE: HVACMode.OFF},
        blocking=True,
    )
    assert hass.states.get(entity_id) == snapshot


async def test_errors(
    hass: HomeAssistant,
    config_entry_mock: MockConfigEntry,
    teslemetry_vehicle_specific_mock,
) -> None:
    """Tests API errors are handled."""

    config_entry_mock.add_to_hass(hass)

    # Test setting climate on with unknown error
    entity_id = "climate.test_climate"
    teslemetry_vehicle_specific_mock.set_temps.side_effect = TeslaFleetError
    await hass.services.async_call(
        CLIMATE_DOMAIN,
        SERVICE_TURN_ON,
        {ATTR_ENTITY_ID: [entity_id]},
        blocking=True,
    )
