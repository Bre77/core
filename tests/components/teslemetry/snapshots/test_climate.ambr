# serializer version: 1
# name: test_climate.1
  None
# ---
# name: test_climate[all]
  0
# ---
